index.html

```diff
@@ -0,0 +1,513 @@
+    view[30] = SimCons.vPlanes[2][3];
+    view[31] = SimCons.vPlanes[3][0];
+    view[32] = SimCons.vPlanes[3][1];
+    view[33] = SimCons.vPlanes[3][2];
+    view[34] = SimCons.vPlanes[3][3];
+    g_updateParams = gl.createBuffer();
+    gl.bindBuffer(gl.UNIFORM_BUFFER, g_updateParams);
+    gl.bufferData(gl.UNIFORM_BUFFER, new Uint8Array(data), gl.STATIC_DRAW);
+    gl.bindBufferBase(gl.UNIFORM_BUFFER, 0, g_updateParams);
+
+    //var mView = [
+    //  1, 0, 0, 0,
+    //  0, 1, 0, 0,
+    //  0, 0, 1, 0,
+    //  -g_fMapWidth/2.0, -g_fMapHeight / 2.0, 0, 1
+    //];
+    //var mProjection = mat4.create();
+    //mat4.frustum(mProjection, -g_fMapWidth/2.0, g_fMapWidth/2.0,
+    //  -g_fMapHeight / 2.0, g_fMapHeight / 2.0, 0, 1);
+    //mat4.multiply(g_mViewProjection, mView, mProjection);
+    ////mat4.multiply(g_mViewProjection, mProjection, mView);
+    //console.log(`mView: ${mView}`);
+    //console.log(`mProjection: ${mProjection}`);
+    //console.log(`g_mVidwProjectionL ${g_mViewProjection}`);
+
+}
+
+function initRender()
+{
+    gl.viewport(0, 0, g_canvas.width, g_canvas.height);
+    gl.clearColor(0.0, 0.0, 0.0, 1.0);
+
+    g_shaderStr[VS] = g_shaderStr[VS].replace(/\${g_iNumParticles}/gi, g_iNumParticles);
+    var vertexShader = loadShader(gl.VERTEX_SHADER, g_shaderStr[VS]);
+    var fragmentShader = loadShader(gl.FRAGMENT_SHADER, g_shaderStr[FS]);
+    g_programs[VS] = gl.createProgram();
+    gl.attachShader(g_programs[VS], vertexShader);
+    gl.attachShader(g_programs[VS], fragmentShader);
+    // Bind a_particlePos to attribute 0
+    // Bind a_particleVel to attribute 1
+    // Bind a_pos to attribute 2
+    gl.bindAttribLocation(g_programs[VS], 0, "a_particlePos");
+    gl.bindAttribLocation(g_programs[VS], 1, "a_particleDen");
+    gl.bindAttribLocation(g_programs[VS], 2, "a_pos");
+    // Link the program
+    gl.linkProgram(g_programs[VS]);
+
+    // Load the vertex data
+    gl.bindBuffer(gl.ARRAY_BUFFER, g_modelBuffer);
+    gl.enableVertexAttribArray(2);
+    gl.vertexAttribPointer(2, 2, gl.FLOAT, false, 0, 0);
+}
+
+function initCS() {
+    [CS_DEN_SIM, CS_DEN_SHA, CS_FOR_SIM, CS_FOR_SHA, CS_INT]
+    .forEach((v) => {
+        g_shaderStr[v] = g_shaderStr[v].replace(/\${g_iNumParticles}/gi, g_iNumParticles); 
+        var cs = loadShader(gl.COMPUTE_SHADER, g_shaderStr[v]);
+        g_programs[v] = gl.createProgram();
+        gl.attachShader(g_programs[v], cs);
+        gl.linkProgram(g_programs[v]);
+    });
+}
+
+function computeDensityPass()
+{
+    gl.useProgram(g_currentDen);
+
+    gl.bindBufferBase(gl.SHADER_STORAGE_BUFFER, 1, g_particleBuffers[0]);
+    gl.bindBufferBase(gl.SHADER_STORAGE_BUFFER, 2, g_particleBuffers[1]);
+    gl.dispatchCompute(g_dispatchNum, 1, 1);
+}
+
+function computeForcePass()
+{
+    gl.useProgram(g_currentFor);
+
+    gl.bindBufferBase(gl.SHADER_STORAGE_BUFFER, 1, g_particleBuffers[0]);
+    gl.bindBufferBase(gl.SHADER_STORAGE_BUFFER, 2, g_particleBuffers[1]);
```

Comment:
Above two lines is not needed since `computeDensityPass()` already did the same binding.

Reply:
done